doc: renew Sphinx configuration

Totktonada opened the pull request:
As I see from https://luafun.github.io, the documentation was generated
last time in 2017 using Sphinx 1.4.8.

I made attempt to regenerate the documentation using the recent Sphinx
version (4.3.0) and observed the following regression.

The documentation has code blocks with highlighted lines
(`:emphasize-lines:`). With the new Sphinx version, a background color
for all lines becomes the same as background color for highlighted ones
(and so the highlight becomes invisible). The cause is [this][1] change,
but if we'll apply [this][2] recipe, the problem disappears.

[1]: https://github.com/sphinx-doc/sphinx/pull/7718
[2]: https://github.com/sphinx-doc/sphinx/pull/5282

I found the relevant update of the default sphinx-quickstart
configuration [3] and decided to renew our configuration as precaution
against all problems of this kind.

[3]: https://github.com/sphinx-doc/sphinx/pull/5700

My steps are the following:

```sh
$ cd doc
$ rm conf.py index.rst Makefile make.bat
$ sphinx-quickstart            \
    --no-sep                   \
    --project 'Lua Functional' \
    --author 'Roman Tsisyk'    \
    -v 0.1                     \
    --release 0.1.3            \
    -l en                      \
    --makefile                 \
    --no-batchfile             \
    --use-make-mode
$ git checkout -- index.rst
```

Most of values are deduced from existing configuration, but I changed
several ones:

* `-l en`: otherwise `sphinx-quickstart` asks about this and, since the
  documentation is in English, it looks appropriate.
* `--no-batchfile`: can't imagine why someone will use Windows to build
  this documentation, so I see no reason to store the batch file in the
  repository.
* `--use-make-mode`: this way `Makefile` will just pass execution to
  `sphinx-build`; it looks as modern way to build Sphinx docs and
  reduces amount of boilerplate code in the repository.

After this I made the following manual modifications:

* Restored copyright year (was 2013-2017, generated as 2021, replaced
  manually with 2013-2021).
* Restored HTML theme from 'alabaster' to 'haiku'.
* Restored `html_logo = 'logo.png'` option.
* Replaced `.. module:: fun` with `.. currentmodule:: fun` over the
  documentation sources to eliminate bunch of warnings:

  > WARNING: duplicate object description of fun, other instance in
  > basic, use :noindex: for one of them

  This way Sphinx will not assume that the module is declared several
  times. Instead, all functions will be added to the module, but the
  alphabetical index will not refer to all those pages as descriptions
  of the 'fun' module.

  Added `.. module::fun` to `reference.rst` to link here from the
  alphabetical index.
* Added `html_copy_source = False` option, because the `*.rst` sources
  are not deployed to the luafun.github.io repository (and it seems it
  would be useless).

Proceed with the following steps to build and open the documentation:

```sh
$ cd doc
$ make html
$ xdg-open _build/html/index.html
```

After all those steps the generated documentation looks mostly the same
as before. I'll note the following differences:

* The background color of code blocks becomes white (was grey).
* A font in function declarations now bigger (at least in my Web
  browser). I guess, it is due to [4].

[4]: https://github.com/sphinx-doc/sphinx/commit/9b3b8a49fb0c2c0059577ccd8993fb6ae64deb97

I'm fine with those changes.

There are some spacing and font differences, the alphabetical index is
now with bullets -- all those changes looks quite minor.

I'll update https://luafun.github.io once this patch will land.

ligurio reviewed (changes requested):
Running `cd doc && make html` failed for me:

```sh
sergeyb@pony:~/sources/MRG/luafun/doc$ make html
Running Sphinx v1.8.5
loading translations [en]... done
making output directory...
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 15 source files that are out of date
updating environment: 15 added, 0 changed, 0 removed
reading sources... [100%] under_the_hood                                                                                                                      

Sphinx error:
master file /home/sergeyb/sources/MRG/luafun/doc/contents.rst not found
```

ligurio commented:
> Changed copyright year from 2021 to 2013-2021 (was 2013-2017).

Probably you mean "Changed copyright year from 2013 to 2021 (was 2013-2017)."

ligurio commented:
```
Proceed with the following steps to build and open the documentation:

$ cd doc
$ make html
$ xdg-open _build/html/index.html
```

Could you put these steps to the README too?

Totktonada commented:
> Running `cd doc && make html` failed for me:
>
> ```
> <...>
> Running Sphinx v1.8.5
> <...>
> ```

Please, re-verify on actual Sphinx version. I didn't test old ones (I hope we shouldn't).

> > Changed copyright year from 2021 to 2013-2021 (was 2013-2017).
> 
> Probably you mean "Changed copyright year from 2013 to 2021 (was 2013-2017)."

Nope. Autogenerated `conf.py` had 2021 as the copyright year, I changed it to 2013-2021. Before this commit it was 2013-2017. Will update the commit message.

> ```
> Proceed with the following steps to build and open the documentation:
>
> $ cd doc
> $ make html
> $ xdg-open _build/html/index.html
> ```
>
> Could you put these steps to the README too?

I would prefer another file like `HACKING.md`. Readme is for users and has enough information already. Will add a commit.

ligurio commented:
> Please, re-verify on actual Sphinx version. I didn't test old ones (I hope we shouldn't).

Works as expected with v4.3.0 installed using PIP.

Totktonada commented:
* Rebased at fresh `master`.
* Changed confusing wording around copyright year.

  Was:

  > Changed copyright year from 2021 to 2013-2021 (was 2013-2017).

  Now:

  > Restored copyright year (was 2013-2017, generated as 2021, replaced
  > manually with 2013-2021).
* Added `HACKING.md` with instruction how to build docs. Mentioned the Sphinx version that works for me.

ligurio commented:
> Added HACKING.md with instruction how to build docs. Mentioned the Sphinx version that works for me.

Is it worth to mention this how to in the readme?

ligurio commented on doc/conf.py:
```diff
@@ -1,247 +1,73 @@
-
-# One entry per manual page. List of tuples
-# (source start file, name, description, authors, manual section).
-man_pages = [
-    ('index', 'luafun', u'Lua Functional Documentation',
-     [u'Roman Tsisyk'], 1)
-]
```
You removed manual page generation. Why?

Totktonada replied:
It is not used anywhere AFAIS (I checked packages building). If we'll need it, we'll enable and carefully verify it, but now it does not seems to be needed.

ligurio commented on doc/conf.py:
```diff
@@ -1,247 +1,73 @@
-# -*- coding: utf-8 -*-
```
Why encoding was removed?

Totktonada replied:
Python 2 is gone.

Totktonada commented:
>> Added HACKING.md with instruction how to build docs. Mentioned the Sphinx version that works for me.
>
> Is it worth to mention this how to in the readme?

This filename is quite usual, but I don't mind. Added.

DifferentialOrange commented on HACKING.md:
````diff
@@ -0,0 +1,13 @@
+
+```sh
+$ cd doc
+$ make html
````
I've got a warning for Sphinx 4.3.0
```
/home/georgymoiseev/Development/github/luafun/luafun/doc/reference.rst:4: WARNING: duplicate object description of fun, other instance in under_the_hood, use :noindex: for one of them
```

Totktonada replied:
Cannot reproduce your result. I have Sphinx 4.3.0 installed from the main Gentoo tree. However I tried to install it using pip and there is no difference: still no warning.

Even if we'll look at the files:

```sh
$ grep -R module.*fun reference.rst under_the_hood.rst
reference.rst:.. module:: fun
under_the_hood.rst:.. currentmodule:: fun
```

We use `currentmodule` in `under_the_hood.rst`, not `module` (see the commit message regarding this). Don't know where the directive is documented, but can link the [changelog entry](https://github.com/sphinx-doc/sphinx/blob/17dfa811078205bd415700361e97e945112b89eb/CHANGES.old#L1100-L1101), where it was introduced.

So, it seems, we should have no warning. Or should have warnings regarding all those `currentmodule` directives in 10 files. One warning is strange.

Please, try to remove the `_build` directory entirely and repeat.

DifferentialOrange replied:
> Please, try to remove the _build directory entirely and repeat.
Yeah, it fixed the issue. Seems like it was some local bug.

DifferentialOrange commented on HACKING.md:
````diff
@@ -0,0 +1,13 @@
+```sh
+$ cd doc
+$ make html
+$ xdg-open _build/html/index.html
````
I'm not familiar with Sphinx, so I've tried to compare current version with result of this build

This build:
![image](https://user-images.githubusercontent.com/20455996/143588750-c35c369b-a145-4472-9606-62a12343e881.png)

luafun.io:
![image](https://user-images.githubusercontent.com/20455996/143588793-0b9a1e62-7894-4440-b264-338e114e2001.png)

It looks like Sphinx 4.3.0 (with thir PR changes or without) is added an extra newline added before `Contents`. Not sure if it's something worth noting.

Totktonada replied:
I'll answer here regarding grey code blocks as well. Your comment:

> It is a matter of personal taste, but I think using grey for code blocks as it was before will be a better idea, if this is easy to set up.

I think it is not hard to do: copy the theme and tweak CSS files. Or, maybe, we can even add our CSS rules without copying the whole theme.

However I don't want step out from 'default haiku theme' (at least within this pull request) mainly because the original documentation uses defaults provided by the theme. A kind of being minimalistic in changes.

Another reason is that anything aside default is a thing of taste and if I'll just declare 'that's default' it should be easier to proceed through review process :)

DifferentialOrange replied:
Okay, sound fair

DifferentialOrange commented on doc/Makefile:
```diff
@@ -1,153 +1,20 @@
-# Makefile for Sphinx documentation
+# Minimal makefile for Sphinx documentation
```
AFAIU, we redirect all existing targets to Sphinx, so this new Makefile is the sane as old one in terms of usability. So I think "minimal" addition is not required.

Totktonada replied:
It is generated Makefile, I didn't touch it. Sphinx developers highlighted that it is minimal (I guess as opposite to one generated by `sphinx-quickstart <...> --no-use-make-mode`), so I would prefer to leave it as is.

DifferentialOrange replied:
Okay, no questions then

DifferentialOrange reviewed (commented):
I left a couple of comments. The only one really useful may be the warning one, other two is more like notes.

It is a matter of personal taste, but I think using grey for code blocks as it was before will be a better idea, if this is easy to set up.

Totktonada commented:
> It is a matter of personal taste, but I think using grey for code blocks as it was before will be a better idea, if this is easy to set up.

Answered [above](https://github.com/luafun/luafun/pull/63#discussion_r759690822).

Totktonada commented:
Steps I performed to renew luafun.github.io:

```sh
$ cd "${LUAFUN_REPOSITORY}"
$ git clean -xffd; git submodule foreach --recursive 'git clean -xffd' # just in case
$ cd doc
$ make html
$ cd "${LUAFUN_GITHUB_IO_REPOSITORY}"
$ git checkout -b Totktonada/regen-using-sphinx-4.3.0
# Drop everything except .nojekyll, which looks necessary:
# https://github.blog/2009-12-29-bypassing-jekyll-on-github-pages/
$ rm -r *.* .buildinfo _static
# The .gitignore file contains only _sources/, but we have no such
# directory and purpose of having it ignored is unclear.
$ rm .gitignore
$ cp -r "${LUAFUN_REPOSITORY}"/doc/_build/html/{*,.buildinfo} ./
$ git add .
$ git commit
$ git push origin Totktonada/regen-using-sphinx-4.3.0
# Open a pull request.
```

Result: https://github.com/luafun/luafun.github.io/pull/1.

Totktonada commented:
Now the result is on https://luafun.github.io/
